### PR TITLE
overlord/snapstate: don't re-enable and start disabled services on refresh, etc.

### PR DESCRIPTION
**Pull request by anonymouse64**
This is a reimplementation of https://github.com/snapcore/snapd/pull/7270, with a few smaller PR's included for clarity. The following dependent PR's were broken out to be merged separately and I will rebase/merge this PR as they are merged:
- [x] https://github.com/snapcore/snapd/pull/7427
- [x] ~~https://github.com/snapcore/snapd/pull/7428~~ closed in favor of https://github.com/snapcore/snapd/pull/7437
- [x] https://github.com/snapcore/snapd/pull/7429
- [x] https://github.com/snapcore/snapd/pull/7430
- [x] https://github.com/snapcore/snapd/pull/7597

This is ready for review I think, but I'm open to options to split smaller things out again if it makes review easier. The commit messages should be reasonably accurate, please read them for more details on the changes specifically.

There are a large number of tests here for the different cases, most of these can be considered regression tests even though we only have two bugs, there are more cases that probably would have failed. 

Fixes: https://bugs.launchpad.net/snapd/+bug/1827237
Fixes: https://bugs.launchpad.net/snapd/+bug/1818306

### CODE REVIEW

File: overlord/snapstate/backend/link.go [hunk 1]
```diff
@@ -143,20 +143,57 @@ func (b Backend) StopServices(apps []*snap.AppInfo, reason snap.ServiceStopReaso
...
+	// cleanup if we fail
+	defer func() {
+		if err == nil {
+			return
+		}
 		wrappers.RemoveSnapBinaries(s)
+	}()
+
+	// add the daemons from the snap.yaml
+	if err = wrappers.AddSnapServices(s, progress.Null); err != nil {
```
Comment (pedronis) [inline overlord/snapstate/backend/link.go]: as I tried to explain, I think this one should take disabledApps (possibly empty)  as well instead of enabling here and disabling shortly after with a different helper

**Comment by anonymouse64** [comment]
there are some interesting issues in the logs from my changes I need to investigate

**Comment by anonymouse64** [comment]
This PR may need to be split up again before it's ready to land, it's approaching 1000 lines...

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -1006,6 +1056,32 @@ func writeSeqFile(name string, snapst *SnapState) error {
 	return osutil.AtomicWriteFile(p, b, 0644, 0)
 }
 
+// ComputeMissingDisabledServices returns a list of services that were disabled
+// that are currently missing from the specific snap info (i.e. they were
+// renamed in this snap info), as well as a list of disabled services that are
+// present in this snap info
+// note that the union of the missing and found svc lists should be exactly
+// equal to st.LastActiveDisabledServices
+func ComputeMissingDisabledServices(st *SnapState, info *snap.Info) ([]string, []string, error) {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: in context the code would be a bit more readable I think if this took st.LastDisabledServices explicity instead of snaps

File: overlord/snapstate/handlers.go [hunk 3]
```diff
@@ -1362,15 +1461,42 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
+		defer st.Lock()
+
+		// get the currently disabled services and add them to
+		// snapst.LastActiveDisabledServices because if we completed a successful
+		// doLinkSnap (hence we are in the undo handler), then we already disabled
+		// the services and deleted currently existing services from the state
+		// during doLinkSnap, but now we will need that information again when we go
+		// to link the old version to prevent accidental enabling of disabled
+		// services on a failed revert/refresh
+		disabledServices, err := m.queryDisabledServices(newInfo, NewTaskProgressAdapterUnlocked(t))
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: shouldn't we store the info as some kind of old* value as we do for other things?  if we are undoing how do we know this information isn't messed up in some way?

File: overlord/snapstate/handlers.go (hunk 2 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: Okay, this change would also make this more easily unit testable

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: hmm I guess I went more with the approach of "be very careful when writing to state" so the old variable stuff didn't seem necessary, but I suppose if this information was messed up somehow we should store it that way instead.

File: overlord/snapstate/handlers.go (hunk 2 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: I changed this in the other PR, I think I will wait to merge in those changes until that PR is merged to master to prevent force pushing or lots of extraneous merge commits.

**Comment by anonymouse64** [comment]
> unit test all the points where we assign:

I will add some more unit tests for this over the weekend

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: I'm slightly confused about the old* pattern, as it seems we only use this pattern for specifically `doLinksnap`/`undoLinkSnap` and don't do it for example with `doUnlinkCurrentSnap`/`undoUnlinkCurrentSnap`. 

Should I also use the old* pattern for `doUnlinkCurrentSnap`/`undoUnlinkCurrentSnap`?

**Comment by anonymouse64** [comment]
While adding more unit tests here around the assignment of LastActiveDisabledServices, I have found that actually we cannot undo a `snap disable` change as there is no undo handler defined for unlink-snap:

https://github.com/snapcore/snapd/blob/17d453671d3580411b3ddb3831dec2f7fffec9ac/overlord/snapstate/snapmgr.go#L409

Is this intentional? If so then I can't test undoing disable, but then again I can't see how to make a `snap disable` change fail in the real world so perhaps it is really a change that can't be undone because it can't fail... 
FWIW though all the other tasks (remove-profiles, remove-aliases, and stop-snap-services) all have undo handlers defined, so just adding an `undoUnlinkSnap` would probably be sufficient to support undoing a disable change.

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (pedronis) [reply overlord/snapstate/handlers.go]: I'm not sure I understand the question.  It depends which places change and how LastActiveDisabledServices, we need to do that consistently, that's why as I said unit tests and tests about where that happens are important. Afaict LastActiveDisabledServices is its own old wrt (do/undo)UnlinkCurrentSnap, unless I'm missing something.

**Comment by pedronis** [comment]
> Is this intentional? If so then I can't test undoing disable, but then again I can't see how to make a snap disable change fail in the real world so perhaps it is really a change that can't be undone because it can't fail...


It's partly intentional at least,  but not completely clear. The issue is that it's used essentially in `disable` and `remove` and it's unclear in those cases if bringing back the snap as available makes sense, or it would even be in a good state. I wouldn't worry too much about that in this context.

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: Okay I think I understand your question. 

How I originally had coded it, yes LastActiveDisabledServices was effectively its own old variable, usable for both successful forward ops like install and undo ops. 

In retrospect, I don't think it was necessary to use the old variables in doLinkSnap/undoLinkSnap unless we are worried about state itself getting corrupted. However, I can see how to implement the old variable pattern for (do/undo)UnlinkCurrentSnap, so I can implement it that way to be consistent if you like. Otherwise perhaps I should revert the change doing the old variables for (do/undo)LinkSnap to be the same

**Comment by anonymouse64** [comment]
> I wouldn't worry too much about that in this context.

Okay, that's fine, I just won't be able to add some unit tests around undoing disables.

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (pedronis) [reply overlord/snapstate/handlers.go]: > However, I can see how to implement the old variable pattern for (do/undo)UnlinkCurrentSnap, so I can implement it that way to be consistent if you like.

undoUnlinkCurrentSnap and undoLinkSnap don't need to use consistent patterns, the reason one uses old* and the other not is they actually deal with very different situations. One doesn't do much except toggling Active and the link (etc). The other crosses revision potentially.

Basically while it's expected for undoLinkSnap to use old*, it's should be needed by undoUnlinkCurrentSnap.

File: overlord/snapstate/backend_test.go [hunk 4]
```diff
@@ -798,24 +800,29 @@ func (f *fakeSnappyBackend) CopySnapData(newInfo, oldInfo *snap.Info, p progress
...
 	}
 
+	op := fakeOp{
+		op:   "link-snap",
+		path: info.MountDir(),
+	}
+
+	// only add the services to the op if there's something to add
+	if len(disabledSvcs) != 0 {
+		op.services = disabledSvcs
```
Comment (stolowski) [inline overlord/snapstate/backend_test.go]: It is slightly confusing that we keep disabled services in 'services' member; should we have a new member for them?

File: overlord/snapstate/handlers.go [hunk 5]
```diff
@@ -783,6 +784,30 @@ func (m *SnapManager) undoMountSnap(t *state.Task, _ *tomb.Tomb) error {
 	return m.backend.RemoveSnapDir(snapsup.placeInfo(), otherInstances)
 }
 
+// queryDisabledServices returns a list of all currently disabled snap services
+// in the snap
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
// in the snap.
```

File: overlord/snapstate/handlers.go [hunk 6]
```diff
@@ -783,6 +784,30 @@ func (m *SnapManager) undoMountSnap(t *state.Task, _ *tomb.Tomb) error {
...
+// note this function takes a snap info rather than snapst because there are
+// situations where we want to call this on non-current snap infos, i.e. in the
+// undo handlers, see undoLinkSnap for an example
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
// undo handlers, see undoLinkSnap for an example.
```

File: overlord/snapstate/handlers.go [hunk 7]
```diff
@@ -783,6 +784,30 @@ func (m *SnapManager) undoMountSnap(t *state.Task, _ *tomb.Tomb) error {
...
+
+	disabledSnapSvcs := []string{}
+	// add all disabled services to the list
+	for svc, isEnabled := range snapSvcStates {
+		if !isEnabled {
+			disabledSnapSvcs = append(disabledSnapSvcs, svc)
+		}
+	}
+
+	return disabledSnapSvcs, nil
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]: Should disabledSnapSvcs be sorted to ease testing?

File: overlord/snapstate/handlers.go [hunk 8]
```diff
@@ -798,6 +823,17 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	// add to the disabled services list in snapst services which were disabled
+	// when stop-snap-services ran, for usage across changes like in reverting
+	// and enabling after being disabled
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// and enabling after being disabled.
```

File: overlord/snapstate/handlers.go [hunk 9]
```diff
@@ -798,6 +823,17 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
...
+	// we keep what's already in the list in snapst because that list is
+	// services which were previously present in the snap and disabled, but are
+	// no longer present
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// no longer present.
```

File: overlord/snapstate/handlers.go [hunk 10]
```diff
@@ -853,12 +889,26 @@ func (m *SnapManager) undoUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	// get the services which LinkSnap should disable when generating wrappers,
+	// as well as the services which are not present in this revision, but were
+	// present and disabled in a previous one and as such should be kept inside
+	// snapst for persistent storage
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// snapst for persistent storage.
```

File: overlord/snapstate/handlers.go [hunk 11]
```diff
@@ -1006,6 +1056,32 @@ func writeSeqFile(name string, snapst *SnapState) error {
 	return osutil.AtomicWriteFile(p, b, 0644, 0)
 }
 
+// ComputeMissingDisabledServices returns a list of services that were disabled
+// that are currently missing from the specific snap info (i.e. they were
+// renamed in this snap info), as well as a list of disabled services that are
+// present in this snap info
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
// present in this snap info.
```

File: overlord/snapstate/handlers.go [hunk 12]
```diff
@@ -1006,6 +1056,32 @@ func writeSeqFile(name string, snapst *SnapState) error {
...
+// note that the union of the missing and found svc lists should be exactly
+// equal to st.LastActiveDisabledServices
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
// equal to st.LastActiveDisabledServices.
```

File: overlord/snapstate/handlers.go [hunk 13]
```diff
@@ -1518,9 +1651,41 @@ func (m *SnapManager) stopSnapServices(t *state.Task, _ *tomb.Tomb) error {
...
 	st.Unlock()
+	defer st.Lock()
+
+	// stop the services
 	err = m.backend.StopServices(svcs, stopReason, pb, perfTimings)
+	if err != nil {
+		return err
+	}
+
+	// get the disabled services after we stopped all the services
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// get the disabled services after we stopped all the services.
```

File: overlord/snapstate/handlers.go [hunk 14]
```diff
@@ -1518,9 +1651,41 @@ func (m *SnapManager) stopSnapServices(t *state.Task, _ *tomb.Tomb) error {
...
+	// NOTE: we could probably do this before we stopped all the services (or
+	// later in a different task from this entirely), but the important ordering
+	// for saving the disabled services is that we save the list before we
+	// unlink the snap (and hence destroy systemd's state of what services are
+	// disabled)
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// disabled).
```

File: overlord/snapstate/handlers.go [hunk 15]
```diff
@@ -1518,9 +1651,41 @@ func (m *SnapManager) stopSnapServices(t *state.Task, _ *tomb.Tomb) error {
...
+	// this list is not meant to save what services are disabled at any given
+	// time, specifically just what services are disabled while systemd loses
+	// track of the services because we need to delete and re-generate the
+	// service units
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// service units.
```

File: overlord/snapstate/handlers.go [hunk 16]
```diff
@@ -1539,12 +1704,23 @@ func (m *SnapManager) doUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
-	err = m.backend.UnlinkSnap(info, pb)
+	// do the final unlink
+	err = m.backend.UnlinkSnap(info, NewTaskProgressAdapterLocked(t))
 	if err != nil {
 		return err
 	}
 
+	// add to the disabled services list in snapst services which were disabled
+	// when stop-snap-services ran, for usage across changes like in reverting
+	// and enabling after being disabled
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// and enabling after being disabled.
```

File: overlord/snapstate/handlers.go [hunk 17]
```diff
@@ -1539,12 +1704,23 @@ func (m *SnapManager) doUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
+	// we keep what's already in the list in snapst because that list is
+	// services which were previously present in the snap and disabled, but are
+	// no longer present
```
Comment (stolowski) [inline overlord/snapstate/handlers.go]:
```suggestion
	// no longer present.
```

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: As discussed in SU, we will keep old* for undoLinkSnap, we don't need to do this for undoUnlinkCurrentSnap.

File: tests/main/disabled-svcs-kept-happy/task.yaml [hunk 18]
```diff
@@ -0,0 +1,444 @@
...
+  sed -i \
+    -e s@THEVERSION@3.0@ \
+    -e s@THESVCNAME@svc2@ \
+    -e "s@THEDAEMON@@" \
+    disabled-svcs-kept3/meta/snap.yaml.in
+  mv disabled-svcs-kept3/meta/snap.yaml.in disabled-svcs-kept3/meta/snap.yaml
+  snap pack ./disabled-svcs-kept3
+
+restore: |
+  if [ -e old_retain_num ]; then
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply tests/main/disabled-svcs-kept-happy/task.yaml]: okay, I wasn't sure if the prepare/restore scripts did this so I wanted to try and have the test cleanup after itself as much as possible, but if this is already done then yeah it's redundant

File: tests/main/disabled-svcs-kept-happy/task.yaml [hunk 19]
```diff
@@ -0,0 +1,444 @@
...
+  fi
+
+  echo "Enable the whole snap"
+  snap enable disabled-svcs-kept
+
+  echo "Check that the service is still disabled"
+  snap services disabled-svcs-kept | MATCH "disabled-svcs-kept\\.svc\\s+disabled\\s+inactive"
+
+  echo "Check that state.json still doesn't contain last-active-disabled-services"
+  if [ "$(jq -r '.data.snaps."disabled-svcs-kept" | ."last-active-disabled-services"?' < /var/lib/snapd/state.json)" != "null" ]; then
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply tests/main/disabled-svcs-kept-happy/task.yaml]: sure there probably is significant gain from doing it with helpers

File: tests/main/disabled-svcs-kept-happy/task.yaml [hunk 20]
```diff
@@ -0,0 +1,444 @@
...
+  ############### CASE 2 #################
+  ########################################
+
+  echo "CASE 2"
+
+  echo "Disable a service in the snap"
+  snap stop --disable disabled-svcs-kept.svc
+
+  echo "Check that it was actually disabled"
+  sleep 2
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply tests/main/disabled-svcs-kept-happy/task.yaml]: Sure

File: tests/main/disabled-svcs-kept-happy/task.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply tests/main/disabled-svcs-kept-happy/task.yaml]: Done

File: overlord/snapstate/snapmgr.go [hunk 21]
```diff
@@ -89,6 +89,11 @@ type SnapSetup struct {
 
 	SnapPath string `json:"snap-path,omitempty"`
 
+	// LastActiveDisabledServices is a list of services that were disabled right
+	// before the snap was unlinked to be used for re-disabling those services
+	// right after re-linking a different revision
+	LastActiveDisabledServices []string
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply overlord/snapstate/snapmgr.go]: Good catch yes this should be there I think

File: overlord/snapstate/backend_test.go (hunk 4 above)
Comment (anonymouse64) [reply overlord/snapstate/backend_test.go]: Sure, I'll add them into a different member

Comment (anonymouse64) [reply overlord/snapstate/backend_test.go]: Adding this revealed that my test coverage of the methods on the fakeSnappyBackend were lacking as none of them failed after initially renaming this, so I'm adding a couple more tests that exercise this part, thanks for the suggestion!

File: overlord/snapstate/handlers.go (hunk 7 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: Due to your previous suggestion about renaming one thing, a number of tests I am adding are made more difficult because this isn't sorted, so you were right this is a good idea and I will implement this :smile:

File: overlord/snapstate/backend/link_test.go [hunk 22]
```diff
@@ -80,7 +80,7 @@ apps:
 `
 	info := snaptest.MockSnap(c, yaml, &snap.SideInfo{Revision: snap.R(11)})
 
-	err := s.be.LinkSnap(info, nil, s.perfTimings)
+	err := s.be.LinkSnap(info, nil, nil, s.perfTimings)
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply overlord/snapstate/backend/link_test.go]: Yes, I added a few tests that test this

File: overlord/snapstate/handlers.go (hunk 7 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: This is done

File: overlord/snapstate/backend_test.go (hunk 4 above)
Comment (anonymouse64) [reply overlord/snapstate/backend_test.go]: This should be done now

**Comment by anonymouse64** [comment]
@stolowski mind giving this another review? Thanks

File: overlord/snapstate/handlers.go [hunk 23]
```diff
@@ -783,6 +785,33 @@ func (m *SnapManager) undoMountSnap(t *state.Task, _ *tomb.Tomb) error {
 	return m.backend.RemoveSnapDir(snapsup.placeInfo(), otherInstances)
 }
 
+// queryDisabledServices returns a list of all currently disabled snap services
+// in the snap.
+// note this function takes a snap info rather than snapst because there are
```
Comment (bboozzoo) [inline overlord/snapstate/handlers.go]:
```suggestion
// Note this function takes a snap info rather than snapst because there are
```

File: overlord/snapstate/handlers.go [hunk 24]
```diff
@@ -1153,6 +1234,7 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 	t.Set("old-devmode", oldDevMode)
 	t.Set("old-jailmode", oldJailMode)
 	t.Set("old-classic", oldClassic)
+	t.Set("old-lastactivedisabledservices", svcsToSave)
```
Comment (bboozzoo) [inline overlord/snapstate/handlers.go]: Looking at other key names, we could use `old-last-active-disabled-services`

File: overlord/snapstate/handlers.go [hunk 25]
```diff
@@ -1384,21 +1471,49 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
+		defer st.Lock()
+
+		// get the currently disabled services and add them to
+		// snapst.LastActiveDisabledServices because if we completed a successful
+		// doLinkSnap (hence we are in the undo handler), then we already disabled
+		// the services and deleted currently existing services from the state
+		// during doLinkSnap, but now we will need that information again when we go
+		// to link the old version to prevent accidental enabling of disabled
+		// services on a failed revert/refresh
+		disabledServices, err := m.queryDisabledServices(newInfo, NewTaskProgressAdapterUnlocked(t))
```
Comment (bboozzoo) [inline overlord/snapstate/handlers.go]: This should probably be refactored a bit to make the tracking of st.Unlock()/st.Lock() a bit easier, eg. like so:
```
st.Unlock()
disabledServices, err := m.queryDisabledServices(newInfo, NewTaskProgressAdapterUnlocked(t))
st.Lock()

File: overlord/snapstate/handlers.go [hunk 26]
```diff
@@ -1548,9 +1663,41 @@ func (m *SnapManager) stopSnapServices(t *state.Task, _ *tomb.Tomb) error {
 
 	pb := NewTaskProgressAdapterUnlocked(t)
 	st.Unlock()
+	defer st.Lock()
+
+	// stop the services
 	err = m.backend.StopServices(svcs, stopReason, pb, perfTimings)
```
Comment (bboozzoo) [inline overlord/snapstate/handlers.go]: same comment about st.Unlock()/std.Lock() here

File: overlord/snapstate/handlers.go (hunk 25 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: As we discussed on IRC, what I have is a bit more readable and safer against future refactors, so I will leave this as is for now.

File: overlord/snapstate/handlers.go (hunk 26 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: As we discussed on IRC, what I have is a bit more readable and safer against future refactors, so I will leave this as is for now.

File: overlord/snapstate/handlers.go (hunk 24 above)
Comment (anonymouse64) [reply overlord/snapstate/handlers.go]: Done
